hipamd/src/hip_mrfs.cpp

```diff
@@ -24,208 +28,610 @@
+        return;
+      }
+    }
+
+    // Create the queue and necessary components
+    {
+      std::lock_guard<std::mutex> lock(_queuesMutex);
+
+      // Initialize queue
+      _queues[queueId] = std::queue<std::shared_ptr<MemcpyTask>>();
+
+      // Create a new condition variable with a unique_ptr
+      _queueConditions[queueId] = std::make_unique<std::condition_variable>();
+      _queueStopFlags[queueId] = false;
+    }
 
-    MRFS_LOG("WORKER", "Thread exiting");
+    // Register queue with QuotaManager - do this outside the lock
+    QuotaManager::instance().registerQueue(queueId);
+
+    // Start dedicated worker thread for this queue
+    createQueueWorker(queueId);
   }
 
   public:
-  // Initialize with specified number of threads
-  MemcpyThreadPool(int numThreads) {
-    MRFS_LOG("POOL", "Creating thread pool with " << numThreads << " threads");
-    for (int i = 0; i < numThreads; ++i) {
-      _workers.emplace_back(&MemcpyThreadPool::workerFunction, this);
-    }
+  // Initialize with available device IDs
+  MemcpyThreadPool(const std::vector<int>& deviceIds) : _deviceIds(deviceIds) {
+    // Queues will be created dynamically as needed
   }
 
   // Clean up
   ~MemcpyThreadPool() {
-    MRFS_LOG("POOL", "Shutting down thread pool");
+    // Signal all worker threads to stop
     {
-      std::lock_guard<std::mutex> lock(_mutex);
-      _stop = true;
+      std::lock_guard<std::mutex> lock(_queuesMutex);
+      for (auto& pair : _queueStopFlags) {
+        pair.second = true;
+      }
+    }
+
+    // Notify all condition variables
+    for (auto& pair : _queueConditions) {
+      pair.second->notify_all();
     }
-    _condition.notify_all();
 
-    for (auto& worker : _workers) {
-      if (worker.joinable()) worker.join();
+    // Join all worker threads
+    for (auto& pair : _queueWorkers) {
+      if (pair.second.joinable()) {
+        pair.second.join();
+      }
     }
-    MRFS_LOG("POOL", "Thread pool shutdown complete");
   }
 
   // Add a new memory copy task
-  void addTask(void* dst, const void* src, size_t sizeBytes, hipMemcpyKind kind,
+  bool addTask(void* dst, const void* src, size_t sizeBytes, hipMemcpyKind kind,
                hip::Stream* stream, int deviceId) {
-    auto task = std::make_shared<MemcpyTask>();
-    task->dst = dst;
-    task->src = src;
-    task->sizeBytes = sizeBytes;
-    task->kind = kind;
-    task->stream = stream;
-    task->deviceId = deviceId;
-
-    int taskId = ++_taskCount;
+    // Create queue ID with current process ID
+    QueueId queueId = {deviceId, getpid()};
+
+    // Ensure queue exists with dedicated worker
+    ensureQueueExists(queueId);
+
+    // Define retry parameters
+    const int maxRetries = 20;
+    const auto retryDelay = std::chrono::milliseconds(100);
+    int retryCount = 0;
+
+    while (retryCount < maxRetries) {
+      // Check with scheduler for quota before issuing load
+      if (QuotaManager::instance().checkQuota(queueId, sizeBytes)) {
```

Comment:
Move this checkQuota to be in a worker thread, not here. We should add the task regardless.